README
SYNOPSIS

    In config.yml:

        plugins:
            Showterm:
                stylesheet: /my_showterm.css

    In your app:

        package MyApp;
        use Dancer ':syntax';
    
        use Dancer::Plugin::Showterm;
    
        ...

DESCRIPTION

    This plugin is a Dancer port of the wonderful http://showterm.io, which
    allows terminal screen captures taken via the UNIX tool
    http://man7.org/linux/man-pages/man1/script.1.html|script to be
    replayed in the browser.

    The plugin will intercept any request for files with a .showterm
    extension and will generate an html page that will be able to replay
    the same file, but with the .typescript extension.

    In other words, if you put the file mysession.typescript in the public
    folder of the app, then the url /mysession.showterm will display its
    webified version. The webified version can also be embedded in other
    pages via iframes:

            <iframe src="/mysession.showterm" 
                    width="660" height="360" style="border: 1px solid #444">
            </iframe>
        

CAPTURING THE ACTION

    The cli capture is done using the UNIX utility c<script>. The plugin
    assumes that the captured screen is 80 columns by 24 rows.

        $ script -ttiming
        ... everything you do here will be recorded ...
        ^D  
        $ echo '---' | cat - timing >> typescript
        $ mv typescript /path/to/dancer/app/public/myscreen.typescript

    Note that the c<typescript> file the plugin uses is the concatenation
    of the original produced typescript with its timing file, separated
    with a type dash on a single line.

ADDED ROUTES

 /showterm/*

    The plugin adds the javascript and stylesheets assets required by the
    webified typescript under /showterm. Those are bundled with the plugin
    as shared tarball. If you want to see where this tarball is on your
    filesystem, you can do

        use Dancer;
        use Dancer::Plugin::Showterm;
    
        print Dancer::Plugin::Showterm->instance->assets_dir;

    or

        $ perl -MFile::ShareDir=dist_dir -E'say dist_dir("Dancer-Plugin-Showterm")'

 *.showterm

    Any request for a file with the extension .showterm will be served the
    showterm page, using the same uri with its extension changed to
    .typescript as the script to play.

CONFIGURATION

        plugins:
            Showterm:
                stylesheet: /my_showterm.css

 stylesheet

    If provided, will be added to the showterm page.

SEE ALSO

    http://showterm.io - the original service



Changes
Revision history for Dancer-Plugin-Showterm

{{$NEXT}}
  [ API CHANGES ]

  [ BUG FIXES ]

  [ DOCUMENTATION ]

  [ ENHANCEMENTS ]

  [ NEW FEATURES ]

  [ STATISTICS ]

0.0.2 2015-10-08
  [ DOCUMENTATION ]
    - Fixed doc link. (GH#1, Justin Cook)

  [ STATISTICS ]
    - code churn: 4 files changed, 28 insertions(+), 1 deletion(-)

0.0.1 2015-09-18
  - First version, unleashed on an unsuspecting world.

  [ STATISTICS ]
    - code churn: 2 files changed, 22 insertions(+), 1 deletion(-)


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.036.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

use File::ShareDir::Install;
$File::ShareDir::Install::INCLUDE_DOTFILES = 1;
$File::ShareDir::Install::INCLUDE_DOTDIRS = 1;
install_share dist => "share";


my %WriteMakefileArgs = (
  "ABSTRACT" => "Replay terminal typescript captures",
  "AUTHOR" => "Yanick Champoux <yanick\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::ShareDir::Install" => "0.06"
  },
  "DISTNAME" => "Dancer-Plugin-Showterm",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dancer::Plugin::Showterm",
  "PREREQ_PM" => {
    "Dancer" => 0,
    "Dancer::Plugin" => 0,
    "File::ShareDir::Tarball" => 0,
    "Moo" => 0,
    "MooX::Singleton" => 0,
    "Path::Tiny" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.0.1",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Dancer" => 0,
  "Dancer::Plugin" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::ShareDir::Install" => "0.06",
  "File::ShareDir::Tarball" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Moo" => 0,
  "MooX::Singleton" => 0,
  "Path::Tiny" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}


lib/Dancer/Plugin/Showterm.pm
package Dancer::Plugin::Showterm;
our $AUTHORITY = 'cpan:YANICK';
# ABSTRACT: Replay terminal typescript captures
$Dancer::Plugin::Showterm::VERSION = '0.0.1';

use strict;
use warnings;

use File::ShareDir::Tarball;

use Dancer ':syntax';
use Dancer::Plugin;
use Path::Tiny;

use Moo;
with 'MooX::Singleton';

has assets_dir => (
    is => 'ro',
    lazy => 1,
    default => sub {
        path(
            plugin_setting->{assets_dir} 
                || File::ShareDir::Tarball::dist_dir('Dancer-Plugin-Showterm') 
        );
    },
);

has stylesheet => (
    is => 'ro',
    default => sub {
        plugin_setting->{stylesheet};
    },
);

my $plugin = __PACKAGE__->instance;

get qr/.*\.showterm/ => sub {
    ( my $file = request->path ) =~ s/\.showterm$/\.typescript/;
    my $template = $plugin->assets_dir->child('showterm.html')->slurp;
    $template =~ s/__FILE__/$file/g;

    $template =~ s[(?=</head>)][
        <link rel="stylesheet" href="$_" />
    ] for grep { $_ } $plugin->stylesheet;

    $template;
};

get qr#/showterm/(.*)# => sub {
    my( $path ) = splat;
    send_file $plugin->assets_dir->child($path), system_path => 1;
};

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Dancer::Plugin::Showterm - Replay terminal typescript captures

=head1 VERSION

version 0.0.1

=head1 SYNOPSIS

In F<config.yml>:

    plugins:
        Showterm:
            stylesheet: /my_showterm.css

In your app:

    package MyApp;
    use Dancer ':syntax';

    use Dancer::Plugin::Showterm;

    ...

=head1 DESCRIPTION

This plugin is a L<Dancer> port of the wonderful L<http://showterm.io>, which allows
terminal screen captures taken via the UNIX tool L<http://man7.org/linux/man-pages/man1/script.1.html|script> to be 
replayed in the browser. 

The plugin will intercept any request for files with a F<.showterm> extension and will generate an
html page that will be able to replay the same file, but with the F<.typescript> extension.

In other words, if you put the file F<mysession.typescript> in the F<public> folder of the app, then the
url F</mysession.showterm> will display its webified version. The webified version can also be embedded in other 
pages via iframes:

        <iframe src="/mysession.showterm" 
                width="660" height="360" style="border: 1px solid #444">
        </iframe>

=head1 CAPTURING THE ACTION

The cli capture is done using the UNIX utility c<script>. 
The plugin assumes that the captured screen is 80 columns by 24 rows.

    $ script -ttiming
    ... everything you do here will be recorded ...
    ^D  
    $ echo '---' | cat - timing >> typescript
    $ mv typescript /path/to/dancer/app/public/myscreen.typescript

Note that the c<typescript> file the plugin uses is the concatenation of the
original produced typescript with its timing file, separated with a type dash 
on a single line.

=head1 ADDED ROUTES

=head2 /showterm/*

The plugin adds the javascript and stylesheets assets required by the 
webified typescript under F</showterm>.  Those are bundled with the plugin as shared tarball.
If you want to see where this tarball is on your filesystem, you can do

    use Dancer;
    use Dancer::Plugin::Showterm;

    print Dancer::Plugin::Showterm->instance->assets_dir;

or

    $ perl -MFile::ShareDir=dist_dir -E'say dist_dir("Dancer-Plugin-Showterm")'

=head2 *.showterm

Any request for a file with the extension F<.showterm> will be served the showterm
page, using the same uri with its extension changed to F<.typescript> as the script to play.

=head1 CONFIGURATION

    plugins:
        Showterm:
            stylesheet: /my_showterm.css

=head2 stylesheet

If provided, will be added to the showterm page.

=head1 SEE ALSO

=over

=item L<http://showterm.io> - the original service

=back

=head1 AUTHOR

Yanick Champoux <yanick@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Yanick Champoux.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
